MSC3911: Deduplication - save file with sha256 path

itsoyou opened the pull request:
# Linked Media MSC3911 AP?: Deduplication- file path with sha256 [#3465](https://github.com/famedly/product-management/issues/3465)
Fixes: #3465

Currently we are storing the file based on `media_id`. With msc3911 introduced, there are many copy actions and that creates duplicated media. So now we try to save file with sha256 path and avoid duplication.

There is a feature flag introduce with `use_sha256_paths`. But even with feature flag is on, we still want to the existing media_id path logic function well, and backward compatible. Feature flag also prevents the existing tests failure.

Previously files were saved like this:
```
media/local_content/Ab/cD/...
media/local_thumbnails/Ab/cD/.../32-32-image-png-crop
media/remote_content/{server_name}/Ab/cD/...
```
But with `use_sha256_paths` enabled, media storage structure would be like this:
```
media/eb/f4/...
thumbnails/eb/f4/.../32-32-image-png-crop
```

The logic to save file with sha256 is quite simple. When we save a new file, it has to be sha256 paths, when we want to retrieve the file or delete the file, it should work for both sha256 path and media_id path.

When we save the media with sha256 path, the `FileInfo()` object must include the new field `sha256`.

itsoyou committed:
feat: save file with sha path

jason-famedly commented on synapse/media/media_storage.py:
```diff
@@ -177,6 +177,7 @@ def __init__(
         self.storage_providers = storage_providers
         self._spam_checker_module_callbacks = hs.get_module_api_callbacks().spam_checker
         self.clock = hs.get_clock()
+        self.use_sha256_paths = hs.config.media.use_sha256_paths
```
Hmm. This may be a problem. The `media` config object will not be filled out for any worker that is not enabled for media(remember the `MediaRepository` and what all that problem was about?).

This is the `MediaStorage` class though, and that should only be enabled for that media worker. Ok, I think I just talked myself out of it being a problem :laughing:

jason-famedly replied:
No, I think it'll be ok living where it is. No worker other than the one for handling media should have access to it, since they shouldn't care about that implementation detail

jason-famedly commented on synapse/media/thumbnailer.py:
```diff
@@ -331,6 +332,9 @@ async def respond_local_thumbnail(
             json_response=restrictions_json,
+            sha256=media_info.sha256
+            if self.use_sha256_paths and media_info.sha256
+            else None,
```
If `media_info.sha256` can be a value or `None`, do we need the condition at all(or any of the similar below)? Especially if using the sha256 for the path is guarded against in the function that produces the path, `MediaStorage._file_info_to_path()`(provided the other comment about merging `_file_info_to_sha256_path` into it is considered)?

jason-famedly commented on synapse/media/_base.py:
```diff
@@ -540,6 +540,7 @@ class FileInfo:
     url_cache: bool = False
     # Whether the file is a thumbnail or not.
     thumbnail: Optional[ThumbnailInfo] = None
+    sha256: Optional[str] = None
```
I think there are a couple spots that maybe `FileInfo` is used that need updating. I saw one in the thumbnailer for sure

jason-famedly commented on synapse/media/thumbnailer.py:
```diff
@@ -788,6 +800,7 @@ def _select_thumbnail(
             file_id: The ID of the media that a thumbnail is being requested for.
             url_cache: True if this is from a URL cache.
             server_name: The server name, if this is a remote thumbnail.
+            sha256: The sha256 of the originalmedia that a thumbnail is being requested for.
```
```suggestion
            sha256: The sha256 of the original media that a thumbnail is being requested for.
```

jason-famedly commented on tests/federation/test_federation_media.py:
```diff
@@ -189,6 +200,80 @@ def test_federation_etag(self) -> None:
+            ip_address: str,
+        ) -> Tuple[int, Dict[bytes, List[bytes]]]:
+            output_stream.write(SMALL_PNG)
+            output_stream.flush()
```
Do we need to `close()` the stream too, or does `flush()` handle that for us?

jason-famedly reviewed (changes requested):
Let's look at this part and come back for another round

itsoyou commented on synapse/media/media_storage.py:
```diff
@@ -350,6 +367,25 @@ async def ensure_media_is_in_local_cache(self, file_info: FileInfo) -> str:
+                method=file_info.thumbnail.method,
+            )
+        return self.filepaths.filepath_sha_rel(file_info.sha256)
+
```
Merging the function wouldn't be a problem for saving/storing process, because we are not using the media_id path when we store it. But when we fetch or delete, even though the sha256 path is enabled, we need both media_id path and sha256 path for backward compatibility. so I think we need the function calculates the media_id path regardless of the flag.

jason-famedly replied:
> (Don't forget to bring the conditions where if the path doesn't exist along for that ride)

So if the sha256 path does not exist(but is enabled) it will fall back to the old media id path?

itsoyou replied:
yes. Now I have an idea that I am including the `_file_info_to_path` logic into the `_file_info_to_sha256_path` so that even when the flag is on, I can use both. I will rename it as well.

jason-famedly committed:
some refinements

deletes

adjust some testing

moved this test case to a place that supports a thread pool, and broke
it into several testable components

remove unused stage_into_file() function

Apply feedback about resolve_path_to_media_file()

Lose this TODO and do not recreate the IOBytes object

Put back this new line, avoid merge conflicts

Abstract actual duplicate media deletion into it's own subroutine

Clean up counting how many references there are to a given media object
based on sha256

Update test to clarify some comments about behavior

Fix copypasta comment without update

lint

jason-famedly committed:
Cargo clippy decided this was needed

I guess because of a rust update?

itsoyou committed:
chore: update to use msc3911 config block

itsoyou committed:
chore: minor changes

jason-famedly commented on tests/rest/client/test_media.py:
```diff
@@ -5112,3 +5122,460 @@ def _check_caching(self, path: str) -> None:
+        self.repo = hs.get_media_repository()
+        self.user = self.register_user("user", "pass")
+        self.tok = self.login("user", "pass")
+
```
I'm sorry, I looked at your changes and realized I may have miscommunicated what I meant. I think we can lose the directory creation in `prepare()`. At this stage in the in the bring up of the server and preparation of the test suite, the homeserver is already running and it's too late to inject a different directory, it's up in `make_homeserver()` as that is for setup that occurs before the server is started(that's where `default_config()` gets called too). Actually, do we need to create this directory at all? It should already be created for us

jason-famedly committed:
fix: Use proper context manager to handle closing the file

jason-famedly committed:
fix: correct media restrictions retrieval SQL(was looking at the wrong object)

jason-famedly committed:
fix: correct test assertion that would have never been true(displayname != avatar url)

jason-famedly committed:
chore: Establish isolated media directory infrastructure

jason-famedly committed:
chore: Reduce boiler plate for setting up isolated media paths across the test suite, and force some tests to always use isolated media paths

itsoyou committed:
chores: docstring update

jason-famedly reviewed (approved):
Looks like good stuff
